contracts/staking/contracts/src/libs/LibFixedMath.sol

```diff
@@ -0,0 +1,376 @@
+                / int256(0x0000000000000000000000000043cbaf42a000812488fc5c220ad7b97bf6e99e); // * e ^ -32
+        }
+        // e ^ -16
+        if ((x & int256(0x0000000000000000000000000000000800000000000000000000000000000000)) != 0) {
+            r = r * int256(0x00000000000000000000000000000000000afe10820813d65dfe6a33c07f738f)
+                / int256(0x000000000000000000000000000005d27a9f51c31b7c2f8038212a0574779991); // * e ^ -16
+        }
+        // e ^ -8
+        if ((x & int256(0x0000000000000000000000000000000400000000000000000000000000000000)) != 0) {
+            r = r * int256(0x0000000000000000000000000000000002582ab704279e8efd15e0265855c47a)
+                / int256(0x0000000000000000000000000000001b4c902e273a58678d6d3bfdb93db96d02); // * e ^ -8
+        }
+        // e ^ -4
+        if ((x & int256(0x0000000000000000000000000000000200000000000000000000000000000000)) != 0) {
+            r = r * int256(0x000000000000000000000000000000001152aaa3bf81cb9fdb76eae12d029571)
+                / int256(0x00000000000000000000000000000003b1cc971a9bb5b9867477440d6d157750); // * e ^ -4
+        }
+        // e ^ -2
+        if ((x & int256(0x0000000000000000000000000000000100000000000000000000000000000000)) != 0) {
+            r = r * int256(0x000000000000000000000000000000002f16ac6c59de6f8d5d6f63c1482a7c86)
+                / int256(0x000000000000000000000000000000015bf0a8b1457695355fb8ac404e7a79e3); // * e ^ -2
+        }
+        // e ^ -1
+        if ((x & int256(0x0000000000000000000000000000000080000000000000000000000000000000)) != 0) {
+            r = r * int256(0x000000000000000000000000000000004da2cbf1be5827f9eb3ad1aa9866ebb3)
+                / int256(0x00000000000000000000000000000000d3094c70f034de4b96ff7d5b6f99fcd8); // * e ^ -1
+        }
+        // e ^ -0.5
+        if ((x & int256(0x0000000000000000000000000000000040000000000000000000000000000000)) != 0) {
+            r = r * int256(0x0000000000000000000000000000000063afbe7ab2082ba1a0ae5e4eb1b479dc)
+                / int256(0x00000000000000000000000000000000a45af1e1f40c333b3de1db4dd55f29a7); // * e ^ -0.5
+        }
+        // e ^ -0.25
+        if ((x & int256(0x0000000000000000000000000000000020000000000000000000000000000000)) != 0) {
+            r = r * int256(0x0000000000000000000000000000000070f5a893b608861e1f58934f97aea57d)
+                / int256(0x00000000000000000000000000000000910b022db7ae67ce76b441c27035c6a1); // * e ^ -0.25
+        }
+        // e ^ -0.125
+        if ((x & int256(0x0000000000000000000000000000000010000000000000000000000000000000)) != 0) {
+            r = r * int256(0x00000000000000000000000000000000783eafef1c0a8f3978c7f81824d62ebf)
+                / int256(0x0000000000000000000000000000000088415abbe9a76bead8d00cf112e4d4a8); // * e ^ -0.125
+        }
+    }
+
+    /// @dev Returns the multiplication two numbers, reverting on overflow.
+    function __mul(int256 a, int256 b) private pure returns (int256 c) {
```

Comment:
Since you started working on this we've switched to using `LibSafeMath`, now that it's a library. Could be good to use it in here as well.

Reply:
I wasn't sure if I wanted to inherit the same error types for `SafeMath` in `FixedMath`. if I did do it I would feel compelled to also throw `SafeMath` errors in places like `__add()`, even though it's very different from `SafeMath.add()`. Then there's other `FixedMathError` types that don't make sense at all for `SafeMath`...